ebpf/src/dockertracker.cpp

```diff
@@ -0,0 +1,267 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "dockertracker.h"
+#include "probes/kprobe_group/header.h"
+
+#include <asm/unistd_64.h>
+
+namespace trailofbits {
+namespace {
+const char* getDockerTrackerErrorDescription(DockerTracker::Error error) {
+  switch (error) {
+  case DockerTracker::Error::Success:
+    return "The event was recorded";
+
+  case DockerTracker::Error::Skipped:
+    return "The event was skipped";
+
+  case DockerTracker::Error::InvalidEvent:
+    return "The specified event is not an exec process event";
+
+  case DockerTracker::Error::BrokenEvent:
+    return "The containerd-shim invocation was missing one or more parameters";
+
+  default:
+    return "<Invalid error code specified>";
+  }
+}
+} // namespace
+
+struct DockerTracker::PrivateData final {
+  DockerTrackerContext context;
+};
+
+DockerTracker::DockerTracker() : d(new PrivateData) {}
+
+osquery::Status DockerTracker::create(DockerTrackerRef& object) {
+  try {
+    object.reset();
+
+    auto ptr = new DockerTracker;
+    object.reset(ptr);
+
+    return osquery::Status(0);
+
+  } catch (const osquery::Status& status) {
+    return status;
+
+  } catch (const std::bad_alloc&) {
+    return osquery::Status::failure("Memory allocation failure");
+  }
+}
+
+DockerTracker::~DockerTracker() {}
+
+void DockerTracker::processProbeEvent(ProbeEvent& probe_event) {
+  processProbeEvent(d->context, probe_event);
+}
+
+bool DockerTracker::queryProcessInformation(std::string& container_id,
+                                            pid_t process_id) {
+  return queryProcessInformation(container_id, d->context, process_id);
+}
+
+void DockerTracker::cleanupContext() {
+  cleanupContext(d->context);
+}
+
+DockerTracker::Error DockerTracker::processContainerShimExecEvent(
+    DockerContainerInstance& container_instance,
+    const ProbeEvent& probe_event) {
+  static const std::unordered_set<int> kSystemCallFilter = {__NR_execve,
+                                                            __NR_execveat};
+
+  container_instance = {};
+
+  if (kSystemCallFilter.count(probe_event.function_identifier) == 0) {
+    return Error::InvalidEvent;
+  }
+
+  std::string filename = {};
+  auto status = getProbeEventField(filename, probe_event, "filename");
+  if (!status.ok()) {
+    return Error::Skipped;
+  }
+
+  auto filename_index = filename.find("/containerd-shim");
+  if (filename_index == std::string::npos ||
+      filename_index + 16U != filename.size()) {
+    return Error::Skipped;
+  }
+
+  ProbeEvent::StringList argv;
+  status = getProbeEventField(argv, probe_event, "argv");
+  if (!status.ok()) {
+    return Error::Skipped;
+  }
+
+  const auto& argv_string_list = argv.data;
+
+  auto namespace_name_it =
+      std::find(argv_string_list.begin(), argv_string_list.end(), "-namespace");
+
+  if (namespace_name_it == argv_string_list.end() ||
```

Comment:
I think these `BrokenEvent` cases would benefit from short explanatory comments: these first two look like checks for malformed arguments, but a comment would help clear that up 😄